### PR TITLE
Make KeyManager configuration more flexible

### PR DESCRIPTION
**Pull request by mosesn**
Motivation:

It's not easy to pass multiple Files as a chain.

Modifications:

Instead of adding interminable combinations of arguments for
SslContextBuilder#keyManager, split off a configuration class specially
for SslContextBuilder#keyManager

Result:

SslContextBuilder#keyManager is easy and flexible to configure.
***
I haven't written any tests yet because I wanted to get some feedback about whether this was the right approach or not.  What do folks think?

### CODE REVIEW

File: handler/src/main/java/io/netty/handler/ssl/KeyManagerConfiguration.java [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+public final class KeyManagerConfiguration {
+
+    private PrivateKey key;
+    private X509Certificate[] keyCertChain;
+    private String keyPassword;
+
+    /**
+     * Creates a new instance.
+     */
+    public KeyManagerConfiguration() { }
```
Comment (netkins) [inline handler/src/main/java/io/netty/handler/ssl/KeyManagerConfiguration.java]: MAJOR Add a nested comment explaining why this method is empty, throw an UnsupportedOperationException or complete the implementation. [rule](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1186)

File: handler/src/main/java/io/netty/handler/ssl/KeyManagerConfiguration.java [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+    public KeyManagerConfiguration key(PrivateKey key, String keyPassword) {
+        this.key = key;
+        this.keyPassword = keyPassword;
+        return this;
+    }
+
+    /**
+     * Sets the private key for this host.
+     *
+     * @param key a PKCS#8 private key file in PEM format
```
Comment (Scottmitch) [inline handler/src/main/java/io/netty/handler/ssl/KeyManagerConfiguration.java]: `key` -> `keyFile`

File: handler/src/main/java/io/netty/handler/ssl/KeyManagerConfiguration.java [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+            return key(SslContext.toPrivateKey(keyFile, keyPassword), keyPassword);
+        } catch (Exception e) {
+            throw new IllegalArgumentException("File does not contain valid private key: " + keyFile, e);
+        }
+    }
+
+    /**
+     * Sets the private key for this host.
+     *
+     * @param key a PKCS#8 private key input stream in PEM format
```
Comment (Scottmitch) [inline handler/src/main/java/io/netty/handler/ssl/KeyManagerConfiguration.java]: `key` -> `keyInputStream`

Comment (Scottmitch) [inline handler/src/main/java/io/netty/handler/ssl/KeyManagerConfiguration.java]: general comment ... double check `@param`s to match up with code.

File: handler/src/main/java/io/netty/handler/ssl/KeyManagerConfiguration.java [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+    /**
+     * Sets the X.509 certificate chain.
+     *
+     * @param keyCertChain an X.509 certificate chain
+     */
+    public KeyManagerConfiguration keyCertChain(X509Certificate... keyCertChain) {
+        if (keyCertChain == null || keyCertChain.length == 0) {
+            this.keyCertChain = null;
+        } else {
+            this.keyCertChain = keyCertChain.clone();
```
Comment (Scottmitch) [inline handler/src/main/java/io/netty/handler/ssl/KeyManagerConfiguration.java]: we don't need to clone if we have already created a new intermediate array from one of the other `keyCertChain` methods. Consider adding an internal method which avoids the clone if called internally.

File: handler/src/main/java/io/netty/handler/ssl/KeyManagerConfiguration.java [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+     * @note Should not be invoked before calling {@code validate}
+     */
+    X509Certificate[] getKeyCertChain() {
+        return keyCertChain;
+    }
+
+    /**
+     * Throws an IllegalArgumentException if the configuration is invalid, returns otherwise.
+     */
+    public void validate(boolean forServer) {
```
Comment (Scottmitch) [inline handler/src/main/java/io/netty/handler/ssl/KeyManagerConfiguration.java]: can this be package private for now?

File: handler/src/main/java/io/netty/handler/ssl/SslContextBuilder.java [hunk 6]
```diff
@@ -305,25 +299,25 @@ public SslContextBuilder keyManager(InputStream keyCertChainInputStream, InputSt
...
+            .keyCertChain(keyCertChain);
+
+        return keyManager(config);
+    }
+
+    /**
+     * Identifying certificate for this host. {@code keyCertChain} and {@code key} may
+     * be {@code null} for client contexts, which disables mutual authentication.
+     *
+     * @param key a PKCS#8 private key file
```
Comment (Scottmitch) [inline handler/src/main/java/io/netty/handler/ssl/SslContextBuilder.java]: please update the javadocs for this method.

File: handler/src/main/java/io/netty/handler/ssl/SslContextBuilder.java [hunk 7]
```diff
@@ -437,15 +429,19 @@ public SslContextBuilder enableOcsp(boolean enableOcsp) {
      * responsible for releasing this object, or else native memory may leak.
      */
     public SslContext build() throws SSLException {
+        X509Certificate[] keyCertChain = keyManagerConfig != null ? keyManagerConfig.getKeyCertChain() : null;
```
Comment (Scottmitch) [inline handler/src/main/java/io/netty/handler/ssl/SslContextBuilder.java]: consider checking the conditional a single time:

```java
final X509Certificate[] keyCertChain;
final PrivateKey key;
final String password;
if (keyManagerConfig != null) {
  keyCertChain = keyManagerConfig.getKeyCertChain();
 ..
} else {
  keyCertChain = null;
..
}


```

**Comment by normanmaurer** [comment]
@mosesn can you please address @Scottmitch s comments ?

**Comment by normanmaurer** [comment]
@mosesn ping

**Comment by normanmaurer** [comment]
@mosesn as it seems like you are not interested anymore let me just close this. Please re-open if you are still addressed and once you addressed the comments. Thanks!

**Comment by mosesn** [comment]
@normanmaurer do you mind if we just keep this open for now?  I'm still snowed in at work on h2 stuff, but I will forget to tackle this when I get some free time if it's marked as closed.

**Comment by normanmaurer** [comment]
@mosesn ok...

**Comment by normanmaurer** [comment]
@mosesn ping

**Comment by mosesn** [comment]
still haven't had time for this.  let's mark it as closed and I'll file an internal twitter ticket to follow up.

**Pull request closed without merging**